x-pack/plugins/maps/public/layers/sources/source.ts

```diff
@@ -0,0 +1,195 @@
+  getIndexPatternIds(): string[];
+  getQueryableIndexPatternIds(): string[];
+  getGeoGridPrecision(zoom: number): number;
+  shouldBeIndexed(): boolean;
+  getPreIndexedShape(): Promise<PreIndexedShape | null>;
+  createFieldFormatter(field: IField): Promise<FieldFormatter | null>;
+  loadStylePropsMeta(args: unknown): Promise<unknown>; // todo
+  getValueSuggestions(field: IField, query: string): Promise<string[]>;
+}
+
+export class AbstractSource implements ISource {
+  static isIndexingSource = false;
+
+  readonly _descriptor: SourceDescriptor;
+  readonly _inspectorAdapters?: object;
+
+  constructor(descriptor: SourceDescriptor, inspectorAdapters?: object) {
+    this._descriptor = descriptor;
+    this._inspectorAdapters = inspectorAdapters;
+  }
+
+  destroy(): void {}
+
+  cloneDescriptor(): SourceDescriptor {
+    // @ts-ignore
+    return copyPersistentState(this._descriptor);
+  }
+
+  async supportsFitToBounds(): Promise<boolean> {
+    return true;
+  }
+
+  /**
+   * return list of immutable source properties.
+   * Immutable source properties are properties that can not be edited by the user.
+   */
+  async getImmutableProperties(): Promise<ImmutableSourceProperty[]> {
+    return [];
+  }
+
+  getInspectorAdapters(): object | undefined {
+    return this._inspectorAdapters;
+  }
+
+  createDefaultLayer(): ILayer {
+    throw new Error(`Source#createDefaultLayer not implemented`);
+  }
+
+  async getDisplayName(): Promise<string> {
+    return '';
+  }
+
+  /**
+   * return attribution for this layer as array of objects with url and label property.
+   * e.g. [{ url: 'example.com', label: 'foobar' }]
+   * @return {Promise<null>}
+   */
+  async getAttributions(): Promise<Attribution[]> {
+    return [];
+  }
+
+  isFieldAware(): boolean {
+    return false;
+  }
+
+  isRefreshTimerAware(): boolean {
+    return false;
+  }
+
+  isGeoGridPrecisionAware(): boolean {
+    return false;
+  }
+
+  isQueryAware(): boolean {
+    return false;
+  }
+
+  getFieldNames(): string[] {
+    return [];
+  }
+
+  renderSourceSettingsEditor() {
+    return null;
+  }
+
+  getApplyGlobalQuery(): boolean {
+    return !!this._descriptor.applyGlobalQuery;
+  }
+
+  getIndexPatternIds(): string[] {
+    return [];
+  }
+
+  getQueryableIndexPatternIds(): string[] {
+    return [];
+  }
+
+  getGeoGridPrecision(zoom: number): number {
+    return 0;
+  }
+
+  isJoinable(): boolean {
+    return false;
+  }
+
+  shouldBeIndexed(): boolean {
+    return AbstractSource.isIndexingSource;
+  }
+
+  isESSource(): boolean {
+    return false;
+  }
+
+  // Returns geo_shape indexed_shape context for spatial quering by pre-indexed shapes
+  async getPreIndexedShape(/* properties */): Promise<PreIndexedShape | null> {
+    return null;
+  }
+
+  // Returns function used to format value
+  async createFieldFormatter(field: IField): Promise<FieldFormatter | null> {
+    return null;
+  }
+
+  async loadStylePropsMeta(args: unknown): Promise<unknown> {
```

Comment:
Maybe `loadStylePropsMeta` should be removed from ISource interface and AbstractSource. It is only implemented in ESAbstractSource and only called VectorLayer after checking that source is [source.isESSource()](https://github.com/elastic/kibana/blob/master/x-pack/plugins/maps/public/layers/vector_layer.js#L463)